### ISSUE TITLE
investigate cmake UNITY_BUILD and precompiled headers

### ISSUE THREAD

**Issue by tjhei** [dealii/dealii#10695]
Based on the experiments in ASPECT [1] we should look at UNITY_BUILD [2],[3] and TARGET_PRECOMPILE_HEADERS [4]. This is a CMake 3.15+ features that potentially improves compile times quite a bit.
For example, this is with the current version of ASPECT (298 vs 32 cc files):
```
~/aspect-git/build-nounity $ time ninja -j 16
real    3m25.492s
~/aspect-git/build-unity $ time ninja -j 16
real    0m54.797s
```
so about 3.6x faster builds.

This was quite easy to do in ASPECT and only required a few small changes. We enabled precompiled headers by default (and left unity builds to when you know that you have enough memory).

@tamiko thoughts?

[1] https://github.com/geodynamics/aspect/issues/3585
[2] https://cmake.org/cmake/help/v3.18/prop_tgt/UNITY_BUILD.html
[3] https://onqtam.com/programming/2019-12-20-pch-unity-cmake-3-16/
[4] https://cmake.org/cmake/help/latest/command/target_precompile_headers.html

**Comment by tamiko** [comment]
@tjhei I like prospect of using cmake-internal precompiled header support. I'll have a look over the weekend :-)

Regarding unity build - @drwells went into great length implementing `DEAL_II_UNITY_BUILD` and I think this is working reasonably well - we could try to port this to the cmake-internal facility, though.

**Comment by drwells** [comment]
I'll take a close look at this at some point this coming week.

**Comment by drwells** [comment]
I don't think that we benefit from switching to CMake's own unity build system, for a few reasons:
1. We have to enable unity builds for each target object library. This means that to restore our current behavior (where we do not add files that are known to take up a lot of compilation resources to the list of unity files) we will have to further split each object library into two parts, e.g.,
```cmake
DEAL_II_ADD_LIBRARY(obj_numerics_separate OBJECT ${_separate_src} ${_header} ${_inst})
EXPAND_INSTANTIATIONS(obj_numerics_separate "${_inst}")
DEAL_II_ADD_LIBRARY(obj_numerics_unity OBJECT ${_unity_include_src} ${_header} ${_inst})

FOREACH(_build ${DEAL_II_BUILD_TYPES})
  STRING(TOLOWER ${_build} _build_lowercase)

  SET_TARGET_PROPERTIES(obj_numerics_unity_${_build_lowercase} PROPERTIES
    UNITY_BUILD ON
    UNITY_BUILD_MODE BATCH
    UNITY_BUILD_BATCH_SIZE 8)
  # TODO: this needs explicit instantiation dependency information since the
  # instantations are really part of a different target.
  ADD_DEPENDENCIES(obj_numerics_unity_${_build_lowercase} obj_numerics_separate_inst)
ENDFOREACH()
```
I personally dislike having to do this.
2. CMake doesn't appear to have any mechanism for cleverly grouping files on its own: for example, its first unity file for the numerics sublibrary is just the first eight files listed in `_unity_include_src`. Hence the end result is the same regardless of if we use our unity build or CMake's.
3. This feature is pretty new and switching to it will prevent users with older copies of CMake from using unity builds.

Precompiled headers sound more useful - perhaps @tamiko can comment more on how we can enable that.

**Comment by tjhei** [comment]
> 1. We have to enable unity builds for each target object library.

I don't understand why you need to do that, I think you just pass ``cmake -D CMAKE_UNITY_BUILD=ON``. We need to be careful and disable it for certain files/targets where we know that it doesn't work.

> 2. CMake doesn't appear to have any mechanism for cleverly grouping files

If you need to control it, you can use https://cmake.org/cmake/help/v3.18/prop_tgt/UNITY_BUILD_MODE.html#prop_tgt:UNITY_BUILD_MODE

> 3. This feature is pretty new and switching to it will prevent users with older copies of CMake from using unity builds.

That is correct. I agree with you that we have an option in place that works, so this is not too important.. But implementing a standard method has its uses as well. A user might try to enable this option (of course we could force it to off...).

> Precompiled headers sound more useful

Yes and I would think that this can help us more as we are already at a point where we split .cc files due to memory usage: no reason to try to merge things together. :-)

**Comment by drwells** [comment]
That's a good point - we could get CMake to generate the same unity build system (up to the bin size) by tagging all the `separate_src` files with `SKIP_UNITY_BUILD_INCLUSION`.

We should definitely force disable this option since some files will definitely fail to compile if we try to collect them (such as anything generated with SPLIT_INSTANTIATIONS).
